File: samples/vdsp_op/warpaffine/warpaffine.py
import vaststreamx as vsx
import numpy as np
import cv2
import argparse
import ast
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)
import common.utils as utils

attr = vsx.AttrKey


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/dog.jpg",
        help="input image",
    )
    parser.add_argument("--output_size", help="output size [w,h]", default="[256, 256]")
    parser.add_argument(
        "--matrix",
        help="warp affine matrix, [x0,x1,x2,y0,y1,y2] ",
        default="[0.7890625, -0.611328125, 56.0, 0.611328125, 0.7890625, -416.0]",
    )
    parser.add_argument(
        "--output_file",
        default="./warpaffine_result.jpg",
        help="output image",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    vsx.set_device(args.device_id)
    output_size = ast.literal_eval(args.output_size)
    matrix = ast.literal_eval(args.matrix)

    cv_image = cv2.imread(args.input_file)
    assert cv_image is not None, f"Failed to read {args.input_file}"

    vsx_image = utils.cv_bgr888_to_vsximage(
        cv_image, vsx.ImageFormat.YUV_NV12, args.device_id
    )
    # Build operator
    op = vsx.BuildInOperator(vsx.OpType.SINGLE_OP_WARP_AFFINE)

    op.set_attribute(
        {
            attr.IIMAGE_FORMAT: vsx.ImageType.YUV_NV12,
            attr.IIMAGE_WIDTH: vsx_image.width,
            attr.IIMAGE_HEIGHT: vsx_image.height,
            attr.IIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.IIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.OIMAGE_WIDTH: output_size[0],
            attr.OIMAGE_HEIGHT: output_size[1],
            attr.OIMAGE_WIDTH_PITCH: output_size[0],
            attr.OIMAGE_HEIGHT_PITCH: output_size[1],
            attr.FLAGS: vsx.ImageWarpAffineMode.WARP_AFFINE_MODE_BILINEAR,
            attr.BORDER_VALUE: [114, 114, 114],
            attr.BORDER_MODE: vsx.ImagePaddingType.PADDING_TYPE_CONSTANT,
            attr.M: matrix,
        }
    )
    output_list = op.execute(
        [vsx_image],
        [(vsx.ImageFormat.YUV_NV12, output_size[0], output_size[1])],
    )

    outs_rgb888 = [utils.vsximage_to_cv_bgr888(out) for out in output_list]
    cv2.imwrite(args.output_file, outs_rgb888[0])

    print(f"save result to {args.output_file}")

File: samples/vdsp_op/warpaffine/warpaffine.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/dog.jpg");
  args.add<std::string>("output_file", '\0', "output image", false,
                        "scale_result.jpg");
  args.add<std::string>("output_size", '\0', "output size [w,h]", false,
                        "[320,320]");
  args.add<std::string>(
      "matrix", '\0', "warpaffine matirx, [x0,x1,x2,y0,y1,y2]", false,
      "[0.7890625, -0.611328125, 56.0, 0.611328125, 0.7890625, -416.0]");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  uint32_t device_id = args.get<uint32_t>("device_id");
  auto input_file = args.get<std::string>("input_file");
  std::vector<uint32_t> output_size =
      vsx::ParseVecUint(args.get<std::string>("output_size"));
  std::vector<float> matrix =
      vsx::ParseVecFloat(args.get<std::string>("matrix"));

  vsx::SetDevice(device_id);

  vsx::Image image;
  if (vsx::MakeVsxImage(input_file, image, vsx::YUV_NV12) != 0) {
    std::cout << "Read input image failed: " << input_file << std::endl;
  }
  int iimage_width = image.Width();
  int iimage_height = image.Height();
  int oimage_width = output_size[0];
  int oimage_height = output_size[1];

  auto image_format = image.Format();

  auto op =
      vsx::BuildInOperator(vsx::BuildInOperatorType::kSINGLE_OP_WARP_AFFINE);

  op.SetAttribute<vsx::AttrKey::kIimageFormat>(
      vsx::BuildInOperatorAttrImageType::kYUV_NV12);
  op.SetAttribute<vsx::AttrKey::kIimageWidth>(iimage_width);
  op.SetAttribute<vsx::AttrKey::kIimageHeight>(iimage_height);
  op.SetAttribute<vsx::AttrKey::kIimageWidthPitch>(iimage_width);
  op.SetAttribute<vsx::AttrKey::kIimageHeightPitch>(iimage_height);

  op.SetAttribute<vsx::AttrKey::kOimageWidth>(oimage_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeight>(oimage_height);
  op.SetAttribute<vsx::AttrKey::kOimageWidthPitch>(oimage_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeightPitch>(oimage_height);

  op.SetAttribute<vsx::AttrKey::kFlags>(
      vsx::ImageWarpAffineFlag::kWARP_AFFINE_FLAG_BILINEAR);

  op.SetAttribute<vsx::AttrKey::kBorderMode>(
      vsx::ImagePaddingType::kPADDING_TYPE_CONSTANT);
  op.SetAttribute<vsx::AttrKey::kBorderValue>({114, 114, 114});
  op.SetAttribute<vsx::AttrKey::kM>(matrix);

  // execute
  // copy input to device memory
  auto input_vacc = vsx::Image(image_format, iimage_width, iimage_height,
                               vsx::Context::VACC(device_id));
  input_vacc.CopyFrom(image);

  // make device memory image for output
  auto output_vacc = vsx::Image(image_format, oimage_width, oimage_width,
                                vsx::Context::VACC(device_id));
  std::vector<vsx::Image> outputs{output_vacc};

  // buildin operator execute
  op.Execute({input_vacc}, outputs);

  // copy result to cpu memory
  vsx::Image output_cpu = vsx::Image(output_vacc.Format(), output_vacc.Width(),
                                     output_vacc.Height());
  output_cpu.CopyFrom(output_vacc);

  cv::Mat out_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output_cpu, out_mat);
  cv::imwrite(args.get<std::string>("output_file"), out_mat);
  std::cout << "save result to : " << args.get<std::string>("output_file")
            << std::endl;
  return 0;
}